File: mppt_po.c


#include "mppt_po.h"

// MPPT parameters
#define STEP_SIZE   0.01f   // Duty cycle step size (1%)
#define DUTY_MIN    0.10f    // Minimum duty cycle (10%)
#define DUTY_MAX    0.90f    // Maximum duty cycle (90%)
#define DUTY_INIT   0.50f    // Initial duty cycle (50%)

// Static variables (persistent between calls)
static float V_prev = 0.0f;
static float P_prev = 0.0f;
static float duty_prev = DUTY_INIT;

/**
 * MPPT Perturb & Observe Function
 * @param V_current: Current PV voltage (Volts)
 * @param I_current: Current PV current (Amps)
 * @return: New duty cycle (0.1 to 0.9)
 */
float mppt_po(float V_current, float I_current) {
    float P_current;      // Current power
    float duty_new;       // New duty cycle to return
    
    // Calculate current power
    P_current = V_current * I_current;
    
    // Perturb & Observe algorithm
    if (P_current > P_prev) {
        // Power increased - continue same direction
        if (V_current > V_prev) {
            duty_new = duty_prev + STEP_SIZE;
        } else {
            duty_new = duty_prev - STEP_SIZE;
        }
    } else {
        // Power decreased - reverse direction
        if (V_current > V_prev) {
            duty_new = duty_prev - STEP_SIZE;
        } else {
            duty_new = duty_prev + STEP_SIZE;
        }
    }
    
    // Limit duty cycle to safe range
    if (duty_new < DUTY_MIN) {
        duty_new = DUTY_MIN;
    }
    if (duty_new > DUTY_MAX) {
        duty_new = DUTY_MAX;
    }
    
    // Update persistent values for next call
    V_prev = V_current;
    P_prev = P_current;
    duty_prev = duty_new;
    
    return duty_new;
}

/**
 * Reset MPPT algorithm (optional)
 * Call this when starting new session
 */
void mppt_reset(void) {
    V_prev = 0.0f;
    P_prev = 0.0f;
    duty_prev = DUTY_INIT;
}

File: Matlab/mppt_po.m
function duty = mppt_po(V, I)
% MPPT Perturb & Observe Algorithm
% Inputs: V = PV voltage, I = PV current
% Output: duty = duty cycle for boost converter

% Persistent variables to remember previous values
persistent Vprev Pprev duty_prev

% Initialize on first call
if isempty(Vprev)
    Vprev = 0;        % Previous voltage
    Pprev = 0;        % Previous power
    duty_prev = 0.5;  % Initial duty cycle (50%)
end

% Calculate current power
P = V * I;

% Perturb & Observe logic
if P > Pprev
    % Power increased → continue same direction
    if V > Vprev
        duty = duty_prev + 0.01;  % Increase duty
    else
        duty = duty_prev - 0.01;  % Decrease duty
    end
else
    % Power decreased → reverse direction
    if V > Vprev
        duty = duty_prev - 0.01;  % Decrease duty
    else
        duty = duty_prev + 0.01;  % Increase duty
    end
end

% Limit duty cycle to safe range (10% to 90%)
duty = max(0.1, min(0.9, duty));

% Update persistent variables for next call
Vprev = V;
Pprev = P;
duty_prev = duty;

end